src/App/Fossa/FossaAPIV1.hs

```diff
@@ -580,3 +587,100 @@ resolveProjectDependencies apiOpts locator = fossaReq $ do
+vsiCreateScanEndpoint baseurl = baseVsiUrl baseurl /: "scans"
+
+vsiCreateScan :: (Has (Lift IO) sig m, Has Diagnostics sig m) => ApiOpts -> ProjectRevision -> m VSI.ScanID
+vsiCreateScan apiOpts ProjectRevision{..} = fossaReq $ do
+  (baseUrl, baseOpts) <- useApiOpts apiOpts
+
+  Organization orgId _ <- getOrganization apiOpts
+  let projectID = renderLocator $ Locator "custom" projectName Nothing
+  let reqBody = VSICreateScanRequestBody orgId projectRevision projectID
+
+  (resBody :: VSICreateScanResponseBody) <- responseBody <$> req POST (vsiCreateScanEndpoint baseUrl) (ReqBodyJson reqBody) jsonResponse baseOpts
+  pure $ unVSICreateScanResponseBody resBody
+
+newtype VSIAddFilesToScanRequestBody = VSIAddFilesToScanRequestBody {vsiAddFilesToScanRequestBodyFiles :: Map (Path Rel File) Fingerprint.Combined}
+
+instance ToJSON VSIAddFilesToScanRequestBody where
+  toJSON VSIAddFilesToScanRequestBody{..} = object ["ScanData" .= vsiAddFilesToScanRequestBodyFiles]
+
+vsiAddFilesToScanEndpoint :: Url scheme -> VSI.ScanID -> Url scheme
+vsiAddFilesToScanEndpoint baseurl scanID = baseVsiUrl baseurl /: "scans" /: toText scanID /: "files"
+
+vsiAddFilesToScan :: (Has (Lift IO) sig m, Has Diagnostics sig m) => ApiOpts -> VSI.ScanID -> Map (Path Rel File) Fingerprint.Combined -> m ()
+vsiAddFilesToScan apiOpts scanID files = fossaReq $ do
+  (baseUrl, baseOpts) <- useApiOpts apiOpts
+
+  let opts = baseOpts <> "sync" =: True
+  let body = VSIAddFilesToScanRequestBody files
+  _ <- req POST (vsiAddFilesToScanEndpoint baseUrl scanID) (ReqBodyJson body) ignoreResponse opts
+
+  pure ()
+
+vsiCompleteScanEndpoint :: Url scheme -> VSI.ScanID -> Url scheme
+vsiCompleteScanEndpoint baseurl scanID = baseVsiUrl baseurl /: "scans" /: toText scanID /: "complete"
+
+vsiCompleteScan :: (Has (Lift IO) sig m, Has Diagnostics sig m) => ApiOpts -> VSI.ScanID -> m ()
+vsiCompleteScan apiOpts scanID = fossaReq $ do
+  (baseUrl, baseOpts) <- useApiOpts apiOpts
+  _ <- req POST (vsiCompleteScanEndpoint baseUrl scanID) NoReqBody ignoreResponse baseOpts
+  pure ()
+
+newtype VSIScanAnalysisStatusBody = VSIScanAnalysisStatusBody {unVSIScanAnalysisStatusBody :: VSI.AnalysisStatus}
+
+instance FromJSON VSIScanAnalysisStatusBody where
+  parseJSON = withObject "VSIScanAnalysisStatusBody" $ \obj -> do
+    (status :: Text) <- obj .: "Status"
+    pure . VSIScanAnalysisStatusBody $ VSI.parseAnalysisStatus status
+
+vsiScanAnalysisStatusEndpoint :: Url scheme -> VSI.ScanID -> Url scheme
+vsiScanAnalysisStatusEndpoint baseurl scanID = baseVsiUrl baseurl /: "scans" /: toText scanID /: "status" /: "analysis"
+
+vsiScanAnalysisStatus :: (Has (Lift IO) sig m, Has Diagnostics sig m) => ApiOpts -> VSI.ScanID -> m VSI.AnalysisStatus
+vsiScanAnalysisStatus apiOpts scanID = fossaReq $ do
+  (baseUrl, baseOpts) <- useApiOpts apiOpts
+  (resBody :: VSIScanAnalysisStatusBody) <- responseBody <$> req GET (vsiScanAnalysisStatusEndpoint baseUrl scanID) NoReqBody jsonResponse baseOpts
```

Comment:
Is this type annotation needed?  I think the newtype unwrapping function below hints the type well enough.

Reply:
Oh! I didn't even think of that. Nice point!